### PR TITLE
KOGITO-1378 Periodic job already running is not canceled

### PR DESCRIPTION
**Pull request by tiagodolphine**
A periodic job already running can be canceled properly.

### CODE REVIEW

**Review by cristianonicolai — APPROVED** [review]

**Comment by radtriste** [comment]
Do we have a reproducer to test this ?

**Comment by sutaakar** [comment]
+1
Which of the test changes actually simulate the issue described in JIRA (invoking the job cancel on periodic job)?
IMHO:
- https://github.com/kiegroup/kogito-apps/pull/174/files#diff-6bfd4f14485feb2bf4797ed88129743bR168 simulates submitting of new job when a job with same id already exists
- https://github.com/kiegroup/kogito-apps/pull/174/files#diff-6bfd4f14485feb2bf4797ed88129743bR293 covers cancelling of non periodic scheduled job. Should this be considered as a reproducer for the JIRA?
- https://github.com/kiegroup/kogito-apps/pull/174/files#diff-6bfd4f14485feb2bf4797ed88129743bR303 covers cancellation of not scheduled job.

File: jobs-service/src/test/java/org/kie/kogito/jobs/service/scheduler/BaseTimerJobSchedulerTest.java [hunk 1]
```diff
@@ -165,6 +165,11 @@ private void testExistingJob(boolean expired, JobStatus jobStatus) {
         verify(tested(), expired ? never() : times(1)).doSchedule(delayCaptor.capture(), eq(job));
         verify(jobRepository, expired ? never() : times(1)).save(scheduleCaptor.capture());
 
+        //assert always a scheduled job is canceled (periodic or not)
+        Optional.ofNullable(jobStatus)
+                .filter(SCHEDULED::equals)
+                .ifPresent(s -> verify(tested()).cancel(scheduleCaptorFuture.capture()));
```
Comment (MarianMacik) [inline jobs-service/src/test/java/org/kie/kogito/jobs/service/scheduler/BaseTimerJobSchedulerTest.java]: Value of this captor is never retrieved.

Comment (tiagodolphine) [reply jobs-service/src/test/java/org/kie/kogito/jobs/service/scheduler/BaseTimerJobSchedulerTest.java]: added the missing assertions.

**Comment by tiagodolphine** [comment]
@radtriste @sutaakar I added a new integration test `JobResourceTest#cancelRunningPeriodicJobTest` that basically reproduces the error without the changes on this PR and passes with the modification on this PR.
The scenario is not easy to test without the integration running on vertx because it is asynchronous. Basically, I create a new periodic job, wait until it is running and after that cancel the running job asserting it was removed on the API perspective and on Vertx, the problem was it was not being removed from Vertx scheduling.
Thanks.

**Comment by sutaakar** [comment]
If I understand it correctly the fix for issue described in JIRA is addressed in https://github.com/kiegroup/kogito-apps/pull/174/files#diff-96206ab3bf634f8422aa73b8f060b41dL334-R334
The issue was related to the fact that the reactive stream created in https://github.com/kiegroup/kogito-apps/pull/174/files#diff-96206ab3bf634f8422aa73b8f060b41dL338 wasn't executed.

The issue also affects non periodic jobs which are already scheduled for execution.

@tiagodolphine Can you please confirm/correct me?

**Comment by tiagodolphine** [comment]
Hi @sutaakar, exactly this is what was causing the issue (on my analysis), and you are right, it was affecting periodic jobs and non-periodic (at least looking at the code), but I'll add a new integration test considering a non-periodic job, in this case, we can make sure and confirm it, this running the test against the code without the modifications on this PR.

**Comment by tiagodolphine** [comment]
@sutaakar added one more test to cancel non-periodic, and it has the issue, the same as periodic.

File: jobs-service/src/test/java/org/kie/kogito/jobs/service/resource/JobResourceTest.java [hunk 2]
```diff
@@ -100,25 +107,115 @@ void getAfterCreate() throws Exception {
         final String id = "3";
         final Job job = getJob(id);
         create(jobToJson(job));
-        final ScheduledJob scheduledJob = given()
+        assertGetScheduledJob(id);
+    }
+
+    @Test
+    void executeTest() throws Exception {
```
Comment (sutaakar) [inline jobs-service/src/test/java/org/kie/kogito/jobs/service/resource/JobResourceTest.java:114]: This test seems to be same as getAfterCreate(), it just contains additional assertions, can they be merged together?

**Review by sutaakar — APPROVED** [review]

**Comment by sutaakar** [comment]
Thanks

**Comment by tiagodolphine** [comment]
@MarianMacik all good for you or it is missing anything for the approval?
@cristianonicolai I think this would be an important fix to be on the 0.9.1, wdyt?
Thanks.

**Comment by tiagodolphine** [comment]
@cristianonicolai It would be great to include this in **0.10.0**.

**Comment by cristianonicolai** [comment]
@tiagodolphine I created the followup JIRA https://issues.redhat.com/browse/KOGITO-1941 as tests are randomly failing in Jenkins, will disable test for now.

**Comment by cristianonicolai** [comment]
PR https://github.com/kiegroup/kogito-apps/pull/191

**Pull request merged**